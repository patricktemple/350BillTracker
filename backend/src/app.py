import logging

from flask import Flask
from flask_marshmallow import Marshmallow

from .settings import DATABASE_URL

app = Flask(__name__)
app.config["SQLALCHEMY_DATABASE_URI"] = DATABASE_URL
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False

marshmallow = Marshmallow(app)

logging.basicConfig(
    format="%(asctime)s [%(levelname)s] %(message)s",
    level=logging.INFO,
    datefmt="%Y-%m-%d %H:%M:%S",
)

# TODO: Can simplify this by just importing the views and letting the rest happen
from . import models, cron, views  # noqa: F401 isort:skip
from .bill import models as bill_models  # noqa: F401 isort:skip
from .bill import views as bill_views
from .person import views as person_views
from .sponsorship import views as sponsorship_views
from .user import views as user_views

from .person import models as person_models  # noqa: F401 isort:skip
from .sponsorship import models as sponsorship_models  # noqa: F401 isort:skip
from .user import models as user_models  # noqa: F401 isort:skip
