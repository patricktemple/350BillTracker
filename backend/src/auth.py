from datetime import datetime, timedelta
from functools import wraps

import jwt
from flask import request

from .models import User
from .settings import JWT_SECRET

JWT_ALGORITHM = "HS256"

# Normally we wouldn't want such long tokens. But the data here is not sensitive
# and the auth is meant to provide just a thin layer of protection. Long browser
# sessions are most important.
# TODO: Extend this. This is short just for testing sakes early on.
TOKEN_LIFETIME = timedelta(hours=1)
JWT_AUDIENCE = "350bt"
JWT_ISSUER = "350bt"

# Notes from:
# https://fusionauth.io/learn/expert-advice/tokens/building-a-secure-jwt/
# https://auth0.com/docs/security/tokens/json-web-tokens/json-web-token-claims


def create_jwt(user_id):
    payload = {
        "sub": str(user_id),
        "exp": datetime.utcnow() + TOKEN_LIFETIME,
        "aud": JWT_AUDIENCE,
        "iss": JWT_ISSUER,
    }
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def verify_jwt(token):
    return jwt.decode(
        token, JWT_SECRET, audience=JWT_AUDIENCE, algorithms=[JWT_ALGORITHM]
    )


# TODO: Rewrite this to invert the auth check
def auth_required(view_fn):
    @wraps(view_fn)
    def check_auth_and_run(*args, **kwargs):
        auth = request.headers.get("Authorization")
        if not auth:
            # TODO: HTTP error types in flask?
            raise ValueError("Authorization required")
        if not auth.startswith("JWT "):
            raise ValueError("Expecting JWT auth type")

        jwt = verify_jwt(auth[4:])

        # Make sure the user still exists
        if not User.query.get(jwt["sub"]):
            raise ValueError("User from JWT no longer exists")

        return view_fn(*args, **kwargs)

    return check_auth_and_run
