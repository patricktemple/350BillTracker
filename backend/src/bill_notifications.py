import logging
from dataclasses import dataclass
from typing import List

from botocore.exceptions import ClientError
from flask import render_template
from sqlalchemy.orm import selectinload

from .bill.views import Bill
from .person.models import Legislator
from .ses import send_email
from .user.models import User


@dataclass
class BillDiff:
    """Utility class to track a bill's state before and after a cron run."""

    bill: Bill = None
    old_status: str = None
    added_sponsor_names: List[str] = None
    removed_sponsor_names: List[str] = None


@dataclass
class BillSnapshot:
    """Utility class to track the state of a bill before a cron run may have
    updated it."""

    status: str = None
    sponsor_ids: List[int] = None


def snapshot_bills():
    """Snapshots the state of all bills. Used to calculate the diff produced by
    a cron job run, so that we can send out email notifications of bill status changes."""
    bills = Bill.query.options(selectinload(Bill.sponsorships)).all()

    snapshots_by_bill_id = {}
    for bill in bills:
        sponsor_ids = [s.legislator_id for s in bill.sponsorships]
        snapshot = BillSnapshot(bill.status, sponsor_ids)
        snapshots_by_bill_id[bill.id] = snapshot

    return snapshots_by_bill_id


def _get_sponsor_subject_string(sponsors):
    if len(sponsors) > 1:
        return f"{len(sponsors)} sponsors"

    return f"sponsor {sponsors[0]}"


def _get_bill_update_subject_line(bill_diffs):
    """Get a subject line for the email describing the bill updates. Assumes the diff list
    and the diffs themselves are non-empty."""
    if len(bill_diffs) > 1:
        return f"{len(bill_diffs)} bills were updated"

    diff = bill_diffs[0]

    # TODO: Update to python 3.10 and use structural pattern matching?
    file = diff.bill.file
    if not diff.added_sponsor_names and not diff.removed_sponsor_names:
        return f"{file}'s status changed to {diff.bill.status}"

    if diff.old_status == diff.bill.status:
        if not diff.added_sponsor_names:
            return f"{file} lost {_get_sponsor_subject_string(diff.removed_sponsor_names)}"
        if not diff.removed_sponsor_names:
            return f"{file} gained {_get_sponsor_subject_string(diff.added_sponsor_names)}"
        return f"{file} gained and lost sponsors"

    return f"{diff.bill.file} was updated"


def _convert_bill_diff_to_template_variables(diff):
    """Converts the intermediate BillDiff into the variables used to render
    the email template."""
    status_changed = diff.bill.status != diff.old_status
    if status_changed:
        status_text = f"Status: {diff.old_status} --> {diff.bill.status}"
        status_color = "blue"
    else:
        status_text = f"Status: {diff.bill.status} (unchanged)"
        status_color = "black"

    sponsors_changed = diff.added_sponsor_names or diff.removed_sponsor_names
    if sponsors_changed:
        new_sponsor_count = len(diff.bill.sponsorships)
        old_sponsor_count = (
            new_sponsor_count
            - len(diff.added_sponsor_names or [])
            + len(diff.removed_sponsor_names or [])
        )
        sponsor_color = "blue"

        # List out the names of all changed sponsors
        explanations = []
        if diff.added_sponsor_names:
            explanations.append(
                f"gained {', '.join(diff.added_sponsor_names)}"
            )
        if diff.removed_sponsor_names:
            explanations.append(
                f"lost {', '.join(diff.removed_sponsor_names)}"
            )

        sponsor_text = f"{old_sponsor_count} sponsors --> {new_sponsor_count} sponsors ({', '.join(explanations)})"
    else:
        sponsor_text = f"{len(diff.bill.sponsorships)} sponsors (unchanged)"
        sponsor_color = "black"

    return {
        "file": diff.bill.file,
        "display_name": diff.bill.display_name,
        "status_text": status_text,
        "status_color": status_color,
        "sponsor_text": sponsor_text,
        "sponsor_color": sponsor_color,
    }


def _send_bill_update_emails(bill_diffs):
    bills_for_template = [
        _convert_bill_diff_to_template_variables(d) for d in bill_diffs
    ]

    subject = _get_bill_update_subject_line(bill_diffs)
    body_text = render_template(
        "bill_alerts_email.txt", bills=bills_for_template
    )
    body_html = render_template(
        "bill_alerts_email.html", bills=bills_for_template
    )

    users_to_notify = User.query.filter_by(
        send_bill_update_notifications=True
    ).all()

    for user in users_to_notify:
        logging.info(f"Sending bill update email to {user.email}")
        try:
            send_email(user.email, subject, body_html, body_text)
        except ClientError:
            logging.exception(f"Faild to send email to {user.email}")


def _calculate_bill_diffs(snapshots_by_bill_id):
    """Looks at the before and after states of all bills, and collapses this info
    info a form that's useful for further processing when building emails."""
    bills = Bill.query.options(selectinload("sponsorships.legislator")).all()

    bill_diffs = []
    for bill in bills:
        snapshot = snapshots_by_bill_id[bill.id]

        added_sponsors = []
        prior_sponsor_ids = set(snapshot.sponsor_ids)
        for sponsorship in bill.sponsorships:
            if sponsorship.legislator_id not in prior_sponsor_ids:
                added_sponsors.append(sponsorship.legislator)
            else:
                prior_sponsor_ids.remove(sponsorship.legislator_id)

        if prior_sponsor_ids:
            removed_sponsors = Legislator.query.filter(
                Legislator.id.in_(prior_sponsor_ids)
            ).all()
        else:
            removed_sponsors = []

        if (
            removed_sponsors
            or added_sponsors
            or bill.status != snapshot.status
        ):
            diff = BillDiff()
            diff.bill = bill
            diff.old_status = snapshot.status
            diff.added_sponsor_names = [s.name for s in added_sponsors]
            diff.removed_sponsor_names = [s.name for s in removed_sponsors]
            bill_diffs.append(diff)

    return bill_diffs


def send_bill_update_notifications(
    snapshots_by_bill_id,
):
    bill_diffs = _calculate_bill_diffs(snapshots_by_bill_id)

    if bill_diffs:
        logging.info("Bills were changed in this cron run, sending emails")
        _send_bill_update_emails(bill_diffs)
