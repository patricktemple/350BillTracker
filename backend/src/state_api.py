import logging

import requests

from .bill.models import (
    AssemblyBill,
    Bill,
    SenateBill,
    StateBill,
    StateChamber,
)
from .models import db
from .person.models import AssemblyMember, Person, Senator
from .settings import SENATE_API_TOKEN
from .sponsorship.models import AssemblySponsorship, SenateSponsorship

# API docs: https://legislation.nysenate.gov/static/docs/html/
# See also https://www.nysenate.gov/how-bill-becomes-law

# Test bills for development
# CCIA: https://nyassembly.gov/leg/?term=2021&bn=S04264
CCIA_ASSEMBLY_ID = "A06967"
CCIA_SENATE_ID = "S04264"
CCIA_TERM = "2021"

CURRENT_SESSION_YEAR = "2021"


def senate_get(path: str, **params):
    response = requests.get(
        f"https://legislation.nysenate.gov/api/3/{path}",
        params={**params, "key": SENATE_API_TOKEN},
    )
    response.raise_for_status()
    return response.json()["result"]


# TODO: Dedupe these fuctions!
def _add_senate_sponsorship(bill, sponsor_data, is_lead_sponsor):
    member_id = sponsor_data["memberId"]
    senator = Senator.query.filter_by(state_member_id=member_id).one_or_none()
    if senator:
        sponsorship = SenateSponsorship(
            senator_id=senator.person_id, is_lead_sponsor=is_lead_sponsor
        )
        bill.state_bill.senate_bill.sponsorships.append(sponsorship)
        logging.info(
            f"Added sponsorship for {senator.person.name} to bill {bill.state_bill.senate_bill.base_print_no}"
        )
    else:
        logging.warning(
            f"Did not find {sponsor_data['fullName']}, member_id: {member_id} for sponsorship on bill {bill.state_bill.senate_bill.base_print_no}"
        )


def _add_assembly_sponsorship(bill, sponsor_data, is_lead_sponsor):
    member_id = sponsor_data["memberId"]
    assembly_member = AssemblyMember.query.filter_by(
        state_member_id=member_id
    ).one_or_none()
    if assembly_member:
        # TODO: I can rename the field "assembly_member_id" and then this function can work for both chambers, it's simpler
        # and that can become a mixin.
        sponsorship = AssemblySponsorship(
            assembly_member_id=assembly_member.person_id,
            is_lead_sponsor=is_lead_sponsor,
        )
        bill.state_bill.assembly_bill.sponsorships.append(sponsorship)
        logging.info(
            f"Added sponsorship for {assembly_member.person.name} to bill {bill.state_bill.assembly_bill.base_print_no}"
        )
    else:
        logging.warning(
            f"Did not find {sponsor_data['fullName']}, member_id: {member_id} for sponsorship on bill {bill.state_bill.assembly_bill.base_print_no}"
        )


def _add_chamber_sponsorships(bill, chamber_data, add_sponsorship_function):
    active_amendment = chamber_data["amendments"]["items"][
        chamber_data["activeVersion"]
    ]
    lead_sponsor = chamber_data["sponsor"]["member"]
    add_sponsorship_function(bill, lead_sponsor, True)
    for sponsor in active_amendment["coSponsors"]["items"]:
        # TODO also include lead sponsor, it's its own field on main bill
        add_sponsorship_function(bill, sponsor, False)


def _update_bill_sponsorships(bill):
    pass


def update_all_sponsorships():
    bills = StateBill.query.all()
    for bill in bills:
        _update_bill_sponsorships(bill)


def import_bill(session_year, base_print_no):
    """Looks up a bill in the State API and starts tracking it. In the state
    API, a "bill" represents either a senate or assembly bill, with linkages
    between the two. In our models, both senate and assembly are captured under
    a single StateBill. Thus, this first looks up the requested senate or assembly
    bill, and then it tries to find out the equivalent bill in the other chamber
    and track that too.
    """
    initial_chamber_response = senate_get(
        f"bills/{session_year}/{base_print_no}", view="no_fulltext"
    )

    # TODO: Filter out resolutions?
    bill = Bill(
        type=Bill.BillType.STATE,
        name=initial_chamber_response["title"],
        description=initial_chamber_response["summary"],
    )
    bill.state_bill = StateBill(session_year=session_year)

    active_amendment = initial_chamber_response["amendments"]["items"][
        initial_chamber_response["activeVersion"]
    ]

    initial_chamber = initial_chamber_response["billType"]["chamber"]

    same_as_versions = active_amendment["sameAs"]["items"]
    alternate_chamber_response = None
    if same_as_versions:
        same_as_print_no = same_as_versions[0]["basePrintNo"]
        alternate_chamber_response = senate_get(
            f"bills/{session_year}/{same_as_print_no}", view="no_fulltext"
        )

        if {
            initial_chamber,
            alternate_chamber_response["billType"]["chamber"],
        } != {"SENATE", "ASSEMBLY"}:
            alternate_chamber_response = None
            logging.error(
                f"Bill {session_year}/{base_print_no} had same_as={same_as_print_no}, in same chamber!"
            )

    if initial_chamber == "SENATE":
        senate_data = initial_chamber_response
        assembly_data = alternate_chamber_response
    else:
        senate_data = alternate_chamber_response
        assembly_data = initial_chamber_response

    if assembly_data:
        bill.state_bill.assembly_bill = AssemblyBill(
            status=assembly_data["status"]["statusDesc"],
            base_print_no=assembly_data["basePrintNo"],
            active_version=assembly_data["activeVersion"],
        )
        _add_chamber_sponsorships(
            bill, assembly_data, _add_assembly_sponsorship
        )
    if senate_data:
        bill.state_bill.senate_bill = SenateBill(
            status=senate_data["status"]["statusDesc"],
            base_print_no=senate_data["basePrintNo"],
            active_version=senate_data["activeVersion"],
        )
        _add_chamber_sponsorships(bill, senate_data, _add_senate_sponsorship)

    db.session.add(bill)
    db.session.commit()
    return bill


def add_state_representatives(session_year=CURRENT_SESSION_YEAR):
    """Queries the NY State API to retrieve all senators and assembly members
    in the current session year. If they already exist in the DB, updates
    their contact info."""

    try:
        members = senate_get(f"members/{session_year}?limit=1000&full=true")

        for member in members["items"]:
            member_id = member["memberId"]
            if member["chamber"] == "ASSEMBLY":
                person_type = Person.PersonType.ASSEMBLY_MEMBER
                existing_member = AssemblyMember.query.filter_by(
                    state_member_id=member_id
                ).one_or_none()
            elif member["chamber"] == "SENATE":
                person_type = Person.PersonType.SENATOR
                existing_member = Senator.query.filter_by(
                    state_member_id=member_id
                ).one_or_none()
            else:
                # ???
                pass

            if existing_member:
                existing_member.district = member["districtCode"]
                logging.info(
                    f"Person {existing_member.person.name} already in DB, updating"
                )
                person = existing_member.person
            else:
                logging.info(f"Adding person {member['person']['fullName']}")
                person = Person(type=person_type)
                if person_type == Person.PersonType.ASSEMBLY_MEMBER:
                    person.assembly_member = AssemblyMember(
                        state_member_id=member_id,
                        district=member["districtCode"],
                    )
                else:
                    person.senator = Senator(
                        state_member_id=member_id,
                        district=member["districtCode"],
                    )
                db.session.add(person)

            person.name = member["person"]["fullName"]
            person.title = member["person"]["prefix"]
            person.email = member["person"]["email"]

        db.session.commit()
    except Exception:
        logging.exception(
            "Unhandled exception when adding state representatives"
        )


def _convert_search_results(state_bill):
    input = state_bill["result"]
    result = {
        "type": Bill.BillType.STATE,
        "name": input["title"],
        "description": input["summary"],
        "session_year": input["session"],
        "base_print_no": input["basePrintNo"],
        "active_version": input["activeVersion"],
        "status": input["status"]["statusDesc"],
        "chamber": StateChamber.SENATE
        if input["billType"]["chamber"] == "SENATE"
        else StateChamber.ASSEMBLY,
    }
    return result


def search_bills(code_name, session_year=None):
    """Searches for a bill by print number and session year. Returns all
    matching results.
    """
    terms = [
        f"(basePrintNo:{code_name} OR printNo:{code_name})",
        "billType.resolution:false",
    ]
    if session_year:
        terms.append(f"session:{session_year}")

    response = senate_get("bills/search", term=" AND ".join(terms))
    return [_convert_search_results(item) for item in response["items"]]
