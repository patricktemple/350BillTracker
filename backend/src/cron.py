import logging
from time import sleep

from . import bill_notifications, council_sync, models
from .app import app
from .settings import ENABLE_CRON


@app.cli.command("cron")
def cron_command():
    # TODO: Refactor to improve error handling
    logging.info("Cron job starting")
    while True:
        if ENABLE_CRON:
            try:
                # logging.info("Syncing data...")
                # logging.info("Adding council members")
                # council_sync.add_council_members()

                # logging.info("Updating council member contact info")
                # council_sync.fill_council_person_data_from_api()

                # logging.info("Refreshing council_member_static_data")
                # This doesn't need to run at cron time though! Just once on startup
                # council_sync.fill_council_person_static_data()

                bill_snapshots = bill_notifications.snapshot_bills()

                logging.info("Syncing all bill updates")
                council_sync.sync_bill_updates()

                logging.info("Syncing all bill sponsorships")
                council_sync.update_all_sponsorships()

                logging.info(
                    "Checking if bills have changed, and sending notifications if so"
                )
                bill_notifications.send_bill_update_notifications(
                    bill_snapshots
                )

                logging.info("Cron run complete")
            except Exception as e:
                logging.exception(e)
                logging.error("Unhandled exception during cron run")
                models.db.session.rollback()
        else:
            logging.info("Cron job is disabled, it won't do anything")

        # TODO: Revert this!
        sleep(1 * 60)
