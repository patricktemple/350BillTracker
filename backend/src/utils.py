import logging
from datetime import datetime, timezone

from src.models import db


def now():
    return datetime.utcnow().replace(tzinfo=timezone.utc)


def cron_function(func):
    def impl(*args, **kwargs):
        try:
            func(*args, **kwargs)
        except Exception:
            db.session.rollback()
            logging.exception("Exception thrown during cron function")

    return impl
