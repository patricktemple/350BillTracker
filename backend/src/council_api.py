from datetime import datetime, timezone

import requests

from src.settings import CITY_COUNCIL_API_TOKEN

from .bill.models import Bill
from .utils import now

from typing import List
import logging

# See http://webapi.legistar.com/Help for an overview of resources.


def council_get(path, *, params=None):
    if not params:
        params = {}
    response = requests.get(
        f"https://webapi.legistar.com/v1/nyc/{path}",
        params={"token": CITY_COUNCIL_API_TOKEN, **params},
    )
    response.raise_for_status()
    return response.json()


def date_filter(field, operator, date):
    return f"{field} {operator} datetime'{date.isoformat()}'"


def eq_filter(field, value):
    return f"{field} eq '{value}'"


def make_filter_param(*filters):
    return {"$filter": " and ".join(filters)}


def _convert_matter_to_bill(matter):
    """Converts the City Council's representation of a bill, called Matters,
    into our own format."""
    return {
        "type": Bill.BillType.CITY,
        "name": matter["MatterName"],
        "description": matter["MatterTitle"],
        "city_bill": {
            "file": matter["MatterFile"],
            "council_body": matter["MatterBodyName"],
            "city_bill_id": matter["MatterId"],
            "intro_date": datetime.fromisoformat(
                matter["MatterIntroDate"]
            ).replace(tzinfo=timezone.utc),
            "status": matter["MatterStatusName"],
            "active_version": matter["MatterVersion"],
        },
    }


def lookup_bills(file_name):
    matters = council_get(
        "matters",
        params=make_filter_param(
            # "Introduction" means "bill". Matters can be other things like "Motion".
            eq_filter("MatterTypeName", "Introduction"),
            f"substringof('{file_name}', MatterFile) eq true",
        ),
    )
    return [_convert_matter_to_bill(m) for m in matters]


def lookup_bill(matter_id):
    matter = council_get(
        f"matters/{matter_id}",
    )
    return _convert_matter_to_bill(matter)


def get_bill_sponsors(matter_id, active_version):
    sponsors = council_get(
        f"matters/{matter_id}/sponsors",
    )

    # The sponsorships for various amendments will appear in a flat list, and we only care
    # about the active version.
    return [
        s
        for s in sponsors
        if s["MatterSponsorMatterVersion"] == active_version
    ]


def get_person(person_id):
    return council_get(f"persons/{person_id}")


def get_current_council_members():
    return council_get(
        "officerecords",
        params=make_filter_param(
            eq_filter("OfficeRecordBodyName", "City Council"),
            date_filter("OfficeRecordStartDate", "le", now().date()),
            date_filter("OfficeRecordEndDate", "ge", now().date()),
        ),
    )


# Unify office records
def get_current_committee_memberships(committee_body_ids: List[int]):
    # An alternate way to filter is to check that OfficeRecordTitle is either 'CHAIRPERSON' or 'Committee Member'.
    # However, this way lets us restrict to the committees that exist in the DB, and could flexibly include new
    # titles in the future.
    records = council_get(
        "officerecords",
        params=make_filter_param(
            # eq_filter("OfficeRecordBodyName", "City Council"),
            date_filter("OfficeRecordStartDate", "le", now().date()),
            date_filter("OfficeRecordEndDate", "ge", now().date()),
        ),
    )
    if len(records) >= 1000:
        logging.error(">=1000 office records returned. This may be getting truncated by lack of pagination")


def get_committees():
    bodies = council_get(
        "bodies",
    )

    # It's weird that Committee on Land Use has a type of Land Use instead of committee
    committee_types = { "Committee", "Subcommittee", "Land Use" }

    # For reference, the other active body types in 2022 were:
    # Primary Legislative Body
    # Charter Revision Commission 2019
    # New York City Advisory Commission on Property Tax Reform
    # Democratic Conference of the Council of the City of New York 
    # Manhattan Borough Board 
    # Minority (Republican) Conference of the Council of the City of New York 
    # Withdrawn
    # Special Committee

    return [b for b in bodies if b['BodyActiveFlag'] and b['BodyTypeName'] in committee_types], bodies