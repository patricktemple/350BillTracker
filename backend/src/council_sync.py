import logging
from datetime import datetime, timezone

from requests import HTTPError

from .bill.models import Bill, CityBill
from .council_api import (
    get_bill_sponsors,
    get_current_council_members,
    get_person,
    lookup_bill,
)
from .models import db
from .person.models import CouncilMember, Person
from .sponsorship.models import CitySponsorship
from .static_data import COUNCIL_DATA_BY_LEGISLATOR_ID
from .utils import now


def add_council_members():
    """Adds basic information about each current council member based on their
    OfficeRecord object in the API. This is missing key fields that need
    to be filled in with other sources."""
    members = get_current_council_members()

    for member in members:
        city_council_person_id = member["OfficeRecordPersonId"]

        existing_council_member = CouncilMember.query.filter_by(
            city_council_person_id=city_council_person_id
        ).one_or_none()
        if existing_council_member:
            council_member = existing_council_member
            person = existing_council_member.person
        else:
            person = Person(type=Person.PersonType.COUNCIL_MEMBER)
            person.council_member = CouncilMember()
            council_member = person.council_member
            db.session.add(person)

        person.name = member["OfficeRecordFullName"]
        person.title = "City Council Member"  # TODO
        council_member.city_council_person_id = city_council_person_id
        council_member.term_start = datetime.fromisoformat(
            member["OfficeRecordStartDate"]
        ).replace(tzinfo=timezone.utc)
        council_member.term_end = datetime.fromisoformat(
            member["OfficeRecordEndDate"]
        ).replace(tzinfo=timezone.utc)

    db.session.commit()


def fill_council_person_data_from_api():
    """For all council members in the DB, updates their contact info and other details
    based on the Person API and our own static data.
    """
    council_members = CouncilMember.query.all()

    # TODO make sure the query joined loads the persons
    for council_member in council_members:
        try:
            data = get_person(council_member.city_council_person_id)
            council_member.person.email = data["PersonEmail"]
            council_member.person.phone = data["PersonPhone"]
            council_member.legislative_phone = data["PersonPhone2"]
            council_member.website = data["PersonWWW"]
            # Borough exists here but we prefer the cleaned static data
        except HTTPError:
            logging.exception(
                f"Could not get Person {council_member.city_council_person_id} from API"
            )
            continue

    db.session.commit()


def fill_council_person_static_data():
    council_members = CouncilMember.query.all()

    for council_member in council_members:
        member_data = COUNCIL_DATA_BY_LEGISLATOR_ID.get(
            council_member.city_council_person_id
        )
        if not member_data:
            logging.warning(
                f"Found a legislator without static data: {council_member.city_council_person_id} {council_member.person.name}"
            )
        else:
            council_member.person.twitter = member_data["twitter"]
            council_member.person.party = member_data["party"]

            # Name and borough both exist in the API but the static data has a
            # cleaned-up version.
            council_member.person.name = member_data["name"]
            council_member.borough = member_data["borough"]

    legislator_ids_from_db = set(
        [l.city_council_person_id for l in council_members]
    )
    if diff := set(COUNCIL_DATA_BY_LEGISLATOR_ID.keys()).difference(
        legislator_ids_from_db
    ):
        unmatched_static_data = [
            COUNCIL_DATA_BY_LEGISLATOR_ID[id] for id in diff
        ]
        logging.warning(
            f"Static data has some legislators not in the DB: {unmatched_static_data}"
        )

    db.session.commit()


# Bills ----------------------------------------------------------------------


def _update_bill(bill):
    bill_data = lookup_bill(bill.city_bill.city_bill_id)
    logging.info(
        f"Updating bill {bill.city_bill.city_bill_id} and got {bill_data}"
    )

    bill.name = bill_data["name"]
    for key in bill_data["city_bill"].keys():
        setattr(bill.city_bill, key, bill_data["city_bill"][key])


def sync_bill_updates():
    bills = Bill.query.filter_by(type=Bill.BillType.CITY).all()

    for bill in bills:
        _update_bill(bill)
        db.session.commit()


def update_bill_sponsorships(city_bill, set_added_at=False):
    """
    Updates sponsorships for a given bill:
    1) Deletes any previous sponsorships
    2) Adds all new sponsorships
    3) If new sponsors aren't in the existing legislators (e.g. they're not longer in office),
       ignore them.
    """
    previous_bill_sponsorships = CitySponsorship.query.filter_by(
        bill_id=city_bill.bill_id
    ).all()
    previous_bill_sponsorships_by_city_id = {
        s.council_member.city_council_person_id: s
        for s in previous_bill_sponsorships
    }

    updated_bill_sponsorships = get_bill_sponsors(city_bill.city_bill_id)

    council_members_for_updated_sponsorships = CouncilMember.query.filter(
        CouncilMember.city_council_person_id.in_(
            [s["MatterSponsorNameId"] for s in updated_bill_sponsorships]
        )
    ).all()
    council_members_for_updated_sponsorships_by_city_id = {
        c.city_council_person_id: c
        for c in council_members_for_updated_sponsorships
    }

    # THIS is super confusing, rewrite it and simplify
    for sponsorship in updated_bill_sponsorships:
        council_member_person_id = sponsorship["MatterSponsorNameId"]

        if (
            council_member_person_id
            not in council_members_for_updated_sponsorships_by_city_id
        ):
            # Can't insert the sponsorship without its foreign key object
            # TODO: Instead, insert a stub for them or something
            logging.warning(
                f"Did not find legislator {council_member_person_id} in db, ignoring..."
            )
            continue

        sponsorship = previous_bill_sponsorships_by_city_id.get(
            council_member_person_id
        )
        if sponsorship:
            # Remove sponsors from this set until we're left with only those
            # sponsorships that were rescinded recently.
            del previous_bill_sponsorships_by_city_id[council_member_person_id]
        else:
            sponsorship = CitySponsorship(
                bill_id=city_bill.bill_id,
                council_member_id=council_members_for_updated_sponsorships_by_city_id[
                    council_member_person_id
                ].person_id,
                sponsor_sequence=sponsorship["MatterSponsorSequence"],
            )
            if set_added_at:
                sponsorship.added_at = now()

            db.session.add(sponsorship)

    for lost_sponsor in previous_bill_sponsorships_by_city_id.values():
        db.session.delete(lost_sponsor)


def update_all_sponsorships():
    bills = Bill.query.all()
    for bill in bills:
        logging.info(f"Updating sponsorships for bill {bill.id} {bill.name}")
        update_bill_sponsorships(bill.city_bill, set_added_at=True)
        db.session.commit()
