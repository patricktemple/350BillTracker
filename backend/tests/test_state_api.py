from unittest.mock import patch

import responses

from src.state_api import add_state_representatives

from .utils import get_response_data


@responses.activate
def test_import_state_reps(client, senator, assembly_member, snapshot):
    responses.add(
        responses.GET,
        url="https://legislation.nysenate.gov/api/3/members/2021?limit=1000&full=true&key=fake_key",
        json={
            "result": {
                "items": [
                    {
                        "memberId": 1,
                        "chamber": "SENATE",
                        "districtCode": 1,
                        "incumbent": True,
                        "person": {
                            "fullName": "New senator",
                            "prefix": "Senator",
                            "email": "a@example.com",
                        },
                    },
                    {
                        "memberId": senator.senator.state_member_id,
                        "chamber": "SENATE",
                        "districtCode": 2,
                        "incumbent": True,
                        "person": {
                            "fullName": "Existing senator",
                            "prefix": "Senator",
                            "email": "b@example.com",
                        },
                    },
                    {
                        "memberId": 1,
                        "chamber": "ASSEMBLY",
                        "districtCode": 3,
                        "incumbent": True,
                        "person": {
                            "fullName": "New assembly member",
                            "prefix": "Assembly member",
                            "email": "c@example.com",
                        },
                    },
                    {
                        "memberId": assembly_member.assembly_member.state_member_id,
                        "chamber": "ASSEMBLY",
                        "districtCode": 4,
                        "incumbent": True,
                        "person": {
                            "fullName": "Existing assembly member",
                            "prefix": "Assembly",
                            "email": "d@example.com",
                        },
                    },
                    {
                        "memberId": 5,
                        "chamber": "ASSEMBLY",
                        "districtCode": 5,
                        "incumbent": False,
                        "person": {
                            "fullName": "New member out of office",
                            "prefix": "Assembly",
                            "email": "d@example.com",
                        },
                    },
                ]
            }
        },
    )
    add_state_representatives()

    response = client.get("/api/persons")

    # Get rid of random UUIDs for snapshot
    response_data = get_response_data(response)
    for item in response_data:
        del item["id"]

    assert response_data == snapshot
