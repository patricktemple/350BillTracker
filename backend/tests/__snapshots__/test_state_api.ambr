# name: test_import_state_reps
  <class 'list'> [
    <class 'dict'> {
      'assemblyMember': <class 'dict'> {
        'district': 3,
        'website': 'https://nyassembly.gov/mem/?ad=3',
      },
      'councilMember': None,
      'email': 'c@example.com',
      'name': 'New assembly member',
      'notes': None,
      'party': None,
      'phone': None,
      'senator': None,
      'title': 'Assembly member',
      'twitter': None,
      'type': 'ASSEMBLY_MEMBER',
    },
    <class 'dict'> {
      'assemblyMember': None,
      'councilMember': None,
      'email': 'a@example.com',
      'name': 'New senator',
      'notes': None,
      'party': None,
      'phone': None,
      'senator': <class 'dict'> {
        'district': 1,
        'website': 'https://www.nysenate.gov/district/1',
      },
      'title': 'Senator',
      'twitter': None,
      'type': 'SENATOR',
    },
    <class 'dict'> {
      'assemblyMember': <class 'dict'> {
        'district': 4,
        'website': 'https://nyassembly.gov/mem/?ad=4',
      },
      'councilMember': None,
      'email': 'me@assembly.com',
      'name': 'assemblymember name',
      'notes': None,
      'party': 'D',
      'phone': '1-555-555-5555',
      'senator': None,
      'title': 'Assemblymember',
      'twitter': 'theassembly',
      'type': 'ASSEMBLY_MEMBER',
    },
    <class 'dict'> {
      'assemblyMember': None,
      'councilMember': None,
      'email': 'me@senate.com',
      'name': 'senator name',
      'notes': None,
      'party': 'D',
      'phone': '1-555-555-5555',
      'senator': <class 'dict'> {
        'district': 2,
        'website': 'https://www.nysenate.gov/district/2',
      },
      'title': 'Senator',
      'twitter': 'thesenateguy',
      'type': 'SENATOR',
    },
  ]
---
